plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java

```diff
@@ -2796,8 +2800,18 @@ public int calculateCpuShares(VirtualMachineTO vmTO) {
 
         if (hostCpuMaxCapacity > 0) {
```

Comment:
@phsm 
I quickly checked the codebase

it looks like for cgroup V1, the hostCpuMaxCapacity is 0.

https://github.com/apache/cloudstack/blob/81e052cfebf5bcdb09e1922393a1753eb6707dae/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java#L3773-L3777

so this pr might not work

Reply:
You are correct: the Cloudstack agent determines that the system uses cgroup v2 correctly.

However, the older libvirt applies the cgroupv1 range as the restriction regardless to which cgroup version it uses.
For example, the libvirt code of version 8.0.0 which is used in Ubuntu 22.04: https://github.com/libvirt/libvirt/blob/v8.0.0/src/conf/domain_validate.c#L1536-L1538 . As you see there, the function just checks the value to be within range, without consideration which cgroup version is in action. 

So technically this PR does not fix a bug in Cloudstack - it adjust Cloudstack to be compatible with the buggy behavior of older Libvirt.

Reply:
thanks for the explanation, it makese sense.

I think more changes are needed.
the checks will not be performed for cgroup V1, because hostCpuMaxCapacity is 0, therefore line 2801
```
if (hostCpuMaxCapacity > 0) {
```
it looks like the checks need to be added in `else`

Reply:
Updated it:
Eliminated the premature return inside `if (hostCpuMaxCapacity > 0) {`, so the return value will be always checked against boundaries before return.